pgpool-II/doc/src/sgml/man1/pg_md5.1
'\" t
.\"     Title: pg_md5
.\"    Author: The Pgpool Global Development Group
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 2023
.\"    Manual: pgpool-II 4.7.1 Documentation
.\"    Source: pgpool-II 4.7.1
.\"  Language: English
.\"
.TH "PG_MD5" "1" "2023" "pgpool-II 4.7.1" "pgpool-II 4.7.1 Documentation"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
pg_md5 \- produces encrypted password in md5
.SH "SYNOPSIS"
.HP \w'\fBpg_md5\fR\ 'u
\fBpg_md5\fR [\fIoption\fR...] \fI\-p\fR
.HP \w'\fBpg_md5\fR\ 'u
\fBpg_md5\fR [\fIoption\fR...] \fIpassword\fR
.SH "DESCRIPTION"
.PP
\fBpg_md5\fR
produces encrypted password in md5\&.
.SH "OPTIONS"
.PP
.PP
\fB\-p\fR
.br
\fB\-\-prompt\fR
.RS 4
Prompt password using standard input\&.
.RE
.PP
\fB\-m\fR
.br
\fB\-\-md5auth\fR
.RS 4
Produce md5 authentication password to authentication file
pool_passwd\&.
.RE
.PP
\fB\-u \fR\fB\fIyour_username\fR\fR
.br
\fB\-\-username=\fR\fB\fIyour_username\fR\fR
.RS 4
When producing a md5 authentication password, create the pool_passwd entry for
your_username\&.
.RE
.PP
\fB\-f \fR\fB\fIconfig_file\fR\fR
.br
\fB\-\-config\-file=\fR\fB\fIconfig_file\fR\fR
.RS 4
Specify the path to the
pgpool\&.conf
configuration file\&.
.RE
.PP
\fB\-i \fR\fB\fIinput_file\fR\fR
.br
\fB\-\-input\-file=\fR\fB\fIinput_file\fR\fR
.RS 4
Specifies file containing user name and password pairs\&.
.RE
.SH "EXAMPLE"
.PP
The following are examples to encrypt your password into md5 hash format for
pcp\&.conf\&.
.sp
.if n \{\
.RS 4
.\}
.nf
$ pg_md5 \-p
password: [your password]
   
.fi
.if n \{\
.RE
.\}
.PP
or
.sp
.if n \{\
.RS 4
.\}
.nf
$ pg_md5 [your password]
acbd18db4cc2f85cedef654fccc4a4d8
  
.fi
.if n \{\
.RE
.\}
.PP
pg_md5
can also be used for adding an entry of user name and
md5
encrypted password to
pool_passwd
authentication file\&.
.sp
.if n \{\
.RS 4
.\}
.nf
$ pg_md5 \-m \-f /path/to/pgpool\&.conf \-u username \-p
password: [your password]

$ cat /path/to/pool_passwd 
username:md55a231fcdb710d73268c4f44283487ba2
    
.fi
.if n \{\
.RE
.\}
.sp
To avoid password prompt or password in command parameter,
pg_md5
can read user name:password pairs from file\&. It will add all user names and
md5
hashed password to
pool_passwd
authentication file\&.
.sp
.if n \{\
.RS 4
.\}
.nf
$ cat users\&.txt
username1:secretpassword1
username2:secretpassword2

$ pg_md5 \-m \-f /path/to/pgpool\&.conf \-i users\&.txt
trying to read username:password pairs from file users\&.txt

$ cat /path/to/pool_passwd
username1:md533314126ba0b187df1e37f5ce6a489a8
username2:md58ae92c6e1d6a48d80e2583fe715e2b36
    
.fi
.if n \{\
.RE
.\}
.sp
To just display the md5 hashed string, not adding an entry to
pool_passwd, pass a string concatenating password and user name\&. For example, if password is "password" and user name is "user", the output would be:
.sp
.if n \{\
.RS 4
.\}
.nf
$ pg_md5 passworduser
4d45974e13472b5a0be3533de4666414
     
.fi
.if n \{\
.RE
.\}
.sp
Please note that the actual entry to be inserted into
pool_passwd
should have "md5" on top of the result string\&. That is: "md54d45974e13472b5a0be3533de4666414"\&.


pgpool-II/pg_md5.c
/* -*-pgsql-c-*- */
/*
 * $Header: /cvsroot/pgpool/pgpool-II/pg_md5.c,v 1.10.2.1 2011/01/13 06:08:16 t-ishii Exp $
 *
 * pgpool: a language independent connection pool server for PostgreSQL
 * written by Tatsuo Ishii
 *
 * Copyright (c) 2003-2010	PgPool Global Development Group
 *
 * Permission to use, copy, modify, and distribute this software and
 * its documentation for any purpose and without fee is hereby
 * granted, provided that the above copyright notice appear in all
 * copies and that both that copyright notice and this permission
 * notice appear in supporting documentation, and that the name of the
 * author not be used in advertising or publicity pertaining to
 * distribution of the software without specific, written prior
 * permission. The author makes no representations about the
 * suitability of this software for any purpose.  It is provided "as
 * is" without express or implied warranty.
 *
 * pg_md5 command main
 *
 */
#include "pool.h"
#include "pool_config.h"
#include "pool_passwd.h"
#include "md5.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <termios.h>
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#else
#include "getopt_long.h"
#endif
#include <pwd.h>
#include <libgen.h>

/* Maximum number of characters allowed for input. */
#define MAX_INPUT_SIZE	32

static void	print_usage(const char prog[], int exit_code);
static void	set_tio_attr(int enable);
static void update_pool_passwd(char *conf_file, char *password);

int
main(int argc, char *argv[])
{
#define PRINT_USAGE(exit_code)	print_usage(argv[0], exit_code)

	char conf_file[POOLMAXPATHLEN+1];
	int opt;
	int optindex;
	bool md5auth = false;
	bool prompt = false;

	static struct option long_options[] = {
		{"help", no_argument, NULL, 'h'},
		{"prompt", no_argument, NULL, 'p'},
		{"md5auth", no_argument, NULL, 'm'},
		{"md5auth", no_argument, NULL, 'm'},
		{"config-file", required_argument, NULL, 'f'},
		{NULL, 0, NULL, 0}
	};

	snprintf(conf_file, sizeof(conf_file), "%s/%s", DEFAULT_CONFIGDIR, POOL_CONF_FILE_NAME);

    while ((opt = getopt_long(argc, argv, "hpmf:", long_options, &optindex)) != -1)
	{
		switch (opt)
		{
			case 'p':    /* prompt for password */
				prompt = true;
				break;

			case 'm':	/* produce md5 authentication password */
				md5auth = true;
				break;

			case 'f':	/* specify configuration file */
				if (!optarg)
				{
					PRINT_USAGE(EXIT_SUCCESS);
				}
				strncpy(conf_file, optarg, sizeof(conf_file));
				break;

			default:
				PRINT_USAGE(EXIT_SUCCESS);
				break;
		}
	}				

	/* Prompt for password. */
	if (prompt)
	{
		char	 md5[MD5_PASSWD_LEN+1];
		char	 buf[MAX_INPUT_SIZE+1];
		int		 len;

		set_tio_attr(1);
		printf("password: ");
		if (!fgets(buf, (MAX_INPUT_SIZE+1), stdin))
		{
			int eno = errno;

			fprintf(stderr, "Couldn't read input from stdin. (fgets(): %s)",
					strerror(eno));

			exit(EXIT_FAILURE);
		}
		set_tio_attr(0);

		/* Remove LF at the end of line, if there is any. */
		len = strlen(buf);
		if (len > 0 && buf[len-1] == '\n')
		{
			buf[len-1] = '\0';
			len--;
		}

		if (md5auth)
		{
			update_pool_passwd(conf_file, buf);
		}
		else
		{
			pool_md5_hash(buf, len, md5);
			printf("%s\n", md5);
		}
	}

	/* Read password from argv. */
	else
	{
		char	md5[POOL_PASSWD_LEN+1];
		int		len;

		if (optind >= argc)
		{
			PRINT_USAGE(EXIT_FAILURE);
		}
			
		len = strlen(argv[optind]);

		if (len > MAX_INPUT_SIZE)
		{
			fprintf(stderr, "Error: Input exceeds maximum password length!\n\n");
			PRINT_USAGE(EXIT_FAILURE);
		}

		if (md5auth)
		{
			update_pool_passwd(conf_file, argv[optind]);
		}
		else
		{
			pool_md5_hash(argv[optind], len, md5);
			printf("%s\n", md5);
		}
	}

	return EXIT_SUCCESS;
}

static void update_pool_passwd(char *conf_file, char *password)
{
	struct passwd *pw;
	char	 md5[MD5_PASSWD_LEN+1];
	char pool_passwd[POOLMAXPATHLEN+1];
	char dirnamebuf[POOLMAXPATHLEN+1];
	char *dirp;

	if (pool_init_config())
	{
		fprintf(stderr, "pool_init_config() failed\n\n");
		exit(EXIT_FAILURE);
	}
	if (pool_get_config(conf_file, INIT_CONFIG))
	{
		fprintf(stderr, "Unable to get configuration. Exiting...");
		exit(EXIT_FAILURE);
	}

	strncpy(dirnamebuf, conf_file, sizeof(dirnamebuf));
	dirp = dirname(dirnamebuf);
	snprintf(pool_passwd, sizeof(pool_passwd), "%s/%s",
			 dirp, pool_config->pool_passwd);
	pool_init_pool_passwd(pool_passwd);

	pw = getpwuid(getuid());
	if (!pw)
	{
		fprintf(stderr, "getpwuid() failed\n\n");
		exit(EXIT_FAILURE);
	}
	pg_md5_encrypt(password, pw->pw_name, strlen(pw->pw_name), md5);
	pool_create_passwdent(pw->pw_name, md5);
	pool_finish_pool_passwd();
}

static void
print_usage(const char prog[], int exit_code)
{
	fprintf(((exit_code == EXIT_SUCCESS) ? stdout : stderr),
			"Usage:\n\
\n\
  %s [OPTIONS]\n\
  %s <PASSWORD>\n\
\n\
  --prompt, -p    Prompt password using standard input.\n\
  --md5auth, -m   Produce md5 authentication password.\n\
  --help, -h      This help menu.\n\
\n\
Warning: At most %d characters are allowed for input.\n\
Warning: Plain password argument is deprecated for security concerns\n\
         and kept for compatibility. Please prefer using password\n\
         prompt.\n",
			prog, prog, MAX_INPUT_SIZE);

	exit(exit_code);
}


static void
set_tio_attr(int set)
{
	struct termios tio;
	static struct termios tio_save;


	if (!isatty(0))
	{
		fprintf(stderr, "stdin is not tty\n");
		exit(EXIT_FAILURE);
	}

	if (set)
	{
		if (tcgetattr(0, &tio) < 0)
		{
			fprintf(stderr, "set_tio_attr(set): tcgetattr failed\n");
			exit(EXIT_FAILURE);
		}

		tio_save = tio;

		tio.c_iflag &= ~(BRKINT|ISTRIP|IXON);
		tio.c_lflag &= ~(ICANON|IEXTEN|ECHO|ECHOE|ECHOK|ECHONL);
		tio.c_cc[VMIN] = 1;
		tio.c_cc[VTIME] = 0;

		if (tcsetattr(0, TCSANOW, &tio) < 0)
		{
			fprintf(stderr, "(set_tio_attr(set): tcsetattr failed\n");
			exit(EXIT_FAILURE);
		}
	}
	else
	{
		if (tcsetattr(0, TCSANOW, &tio_save) < 0)
		{
			fprintf(stderr, "set_tio_attr(reset): tcsetattr failed\n");
			exit(EXIT_FAILURE);
		}
	}
}
